routing/payment_lifecycle.go

```diff
@@ -722,6 +733,75 @@ func (p *paymentLifecycle) sendAttempt(
 	}, nil
 }
 
+// amendFirstHopData is a function that calls the traffic shaper to allow it to
+// add custom records to the outgoing HTLC and also adjust the amount if
+// needed.
+func (p *paymentLifecycle) amendFirstHopData(rt *route.Route) error {
+	// The first hop amount on the route is the full route amount if not
+	// overwritten by the traffic shaper. So we set the initial value now
+	// and potentially overwrite it later.
+	rt.FirstHopAmount = tlv.NewRecordT[tlv.TlvType0](
+		tlv.NewBigSizeT(rt.TotalAmount),
+	)
+
+	// By default, we set the first hop custom records to the initial
+	// value requested by the RPC. The traffic shaper may overwrite this
+	// value.
+	rt.FirstHopWireCustomRecords = p.firstHopCustomRecords
+
+	// extraDataRequest is a helper struct to pass the custom records and
+	// amount back from the traffic shaper.
+	type extraDataRequest struct {
+		customRecords fn.Option[lnwire.CustomRecords]
+
+		amount fn.Option[lnwire.MilliSatoshi]
+	}
+
+	// If a hook exists that may affect our outgoing message, we call it now
+	// and apply its side effects to the UpdateAddHTLC message.
+	result, err := fn.MapOptionZ(
+		p.router.cfg.TrafficShaper,
+		func(ts TlvTrafficShaper) fn.Result[extraDataRequest] {
+			newAmt, newRecords, err := ts.ProduceHtlcExtraData(
```

Comment:
Yeah, that's on purpose. It might be that the custom routing information negotiated with the peer (RFQ) has expired in the meantime. So we should always call into this method so we can error out if the payment cannot succeed anymore.
Also, the amount returned here is kind of static (always just above dust), so it shouldn't change between calls. Or did you mean that it should be persisted as well?